### ISSUE TITLE
UI lag: ChangedValueChecker.hasConflict()

### ISSUE THREAD

**Issue by jukzi** [eclipse-jdt/eclipse.jdt.ui#1185]
up to ~38sec UI lag when pressing STRG+1 at JDT.core's only line that contains a warning in platform workspace: AbstractCommentParser.java:1426

What's going on there?
 
![image](https://github.com/eclipse-jdt/eclipse.jdt.ui/assets/51790620/94967169-b8fb-4179-a09f-39dea4f2b261)
Sampling Results:
UI Thread:
```
org.eclipse.jface.text.quickassist.QuickAssistAssistant$ContentAssistProcessor.computeCompletionProposals ()
...
org.eclipse.jdt.internal.corext.refactoring.util.ChangedValueChecker.hasConflict ()
java.util.concurrent.ThreadPoolExecutor.awaitTermination ()
```
...
![image](https://github.com/eclipse-jdt/eclipse.jdt.ui/assets/51790620/5906ffb3-b122-471b-a16c-3a5b4e83e2ed)
Threadpool:
![image](https://github.com/eclipse-jdt/eclipse.jdt.ui/assets/51790620/5ef4fd3f-005c-4598-9901-176c184935f8)
![image](https://github.com/eclipse-jdt/eclipse.jdt.ui/assets/51790620/c339aa02-3855-4dd3-8fef-4c9b5863f7e2)
no hotspots:
![image](https://github.com/eclipse-jdt/eclipse.jdt.ui/assets/51790620/7db38d9f-7812-4c5f-ba22-16a1d374f4cb)

**Comment by jukzi** [comment]
that code is from https://github.com/eclipse-jdt/eclipse.jdt.ui/pull/439

**Comment by jukzi** [comment]
stacktrace:
```
Thread [main] (Suspended (breakpoint at line 124 in org.eclipse.jdt.internal.corext.refactoring.util.ChangedValueChecker))	
	org.eclipse.jdt.internal.corext.refactoring.util.ChangedValueChecker.hasConflict() line: 124	
	org.eclipse.jdt.internal.corext.refactoring.code.ExtractTempRefactoring.evalStartAndEnd(org.eclipse.jdt.internal.corext.dom.fragments.IASTFragment[], int, java.lang.Integer) line: 1222	
	org.eclipse.jdt.internal.corext.refactoring.code.ExtractTempRefactoring.createAndInsertTempDeclaration() line: 1171	
	org.eclipse.jdt.internal.corext.refactoring.code.ExtractTempRefactoring.createTempDeclaration() line: 1372	
	org.eclipse.jdt.internal.corext.refactoring.code.ExtractTempRefactoring.processSelectedExpression() line: 880	
	org.eclipse.jdt.internal.corext.refactoring.code.ExtractTempRefactoring.doCreateChange(org.eclipse.core.runtime.IProgressMonitor) line: 825	
	org.eclipse.jdt.internal.corext.refactoring.code.ExtractTempRefactoring.checkFinalConditions(org.eclipse.core.runtime.IProgressMonitor) line: 736	
	org.eclipse.jdt.internal.ui.text.correction.QuickAssistProcessor.getExtractVariableProposal(org.eclipse.jdt.ui.text.java.IInvocationContext, boolean, java.util.Collection<org.eclipse.jdt.ui.text.java.correction.ICommandAccess>) line: 580	
	org.eclipse.jdt.internal.ui.text.correction.QuickAssistProcessor.getAssists(org.eclipse.jdt.ui.text.java.IInvocationContext, org.eclipse.jdt.ui.text.java.IProblemLocation[]) line: 387	
	org.eclipse.jdt.internal.ui.text.correction.JavaCorrectionProcessor$SafeAssistCollector.safeRun(org.eclipse.jdt.internal.ui.text.correction.ContributedProcessorDescriptor) line: 404	
	org.eclipse.jdt.internal.ui.text.correction.JavaCorrectionProcessor$SafeAssistCollector(org.eclipse.jdt.internal.ui.text.correction.JavaCorrectionProcessor$SafeCorrectionProcessorAccess).run() line: 341	
	org.eclipse.core.runtime.SafeRunner.run(org.eclipse.core.runtime.ISafeRunnable) line: 47	
	org.eclipse.jdt.internal.ui.text.correction.JavaCorrectionProcessor$SafeAssistCollector(org.eclipse.jdt.internal.ui.text.correction.JavaCorrectionProcessor$SafeCorrectionProcessorAccess).process(org.eclipse.jdt.internal.ui.text.correction.ContributedProcessorDescriptor[]) line: 330	
	org.eclipse.jdt.internal.ui.text.correction.JavaCorrectionProcessor.collectAssists(org.eclipse.jdt.ui.text.java.IInvocationContext, org.eclipse.jdt.ui.text.java.IProblemLocation[], java.util.Collection<org.eclipse.jdt.ui.text.java.IJavaCompletionProposal>) line: 506	
	org.eclipse.jdt.internal.ui.text.correction.JavaCorrectionProcessor.collectProposals(org.eclipse.jdt.ui.text.java.IInvocationContext, org.eclipse.jface.text.source.IAnnotationModel, org.eclipse.jface.text.source.Annotation[], boolean, boolean, java.util.Collection<org.eclipse.jdt.ui.text.java.IJavaCompletionProposal>) line: 291	
	org.eclipse.jdt.internal.ui.text.correction.JavaCorrectionProcessor.computeQuickAssistProposals(org.eclipse.jface.text.quickassist.IQuickAssistInvocationContext) line: 247	
	org.eclipse.jface.text.quickassist.QuickAssistAssistant$ContentAssistProcessor.computeCompletionProposals(org.eclipse.jface.text.ITextViewer, int) line: 71	
	org.eclipse.jface.text.contentassist.ContentAssistant$2.lambda$0(org.eclipse.jface.text.ITextViewer, int, java.util.List, org.eclipse.jface.text.contentassist.IContentAssistProcessor) line: 2064	
	org.eclipse.jface.text.contentassist.ContentAssistant$2$$Lambda.0x00000173d1a09560.accept(java.lang.Object) line: not available	
	java.util.Collections$SingletonSet<E>.forEach(java.util.function.Consumer<? super E>) line: 5125	
	org.eclipse.jface.text.contentassist.ContentAssistant$2.run() line: 2063	
	org.eclipse.core.runtime.SafeRunner.run(org.eclipse.core.runtime.ISafeRunnable) line: 47	
	org.eclipse.jface.text.quickassist.QuickAssistAssistant$QuickAssistAssistantImpl(org.eclipse.jface.text.contentassist.ContentAssistant).computeCompletionProposals(org.eclipse.jface.text.ITextViewer, int) line: 2060	
	org.eclipse.jface.text.contentassist.CompletionProposalPopup.computeProposals(int) line: 577	
	org.eclipse.jface.text.contentassist.CompletionProposalPopup.lambda$0(boolean) line: 507	
	org.eclipse.jface.text.contentassist.CompletionProposalPopup$$Lambda.0x00000173d1a08ec8.run() line: not available	
	org.eclipse.swt.custom.BusyIndicator.showWhile(org.eclipse.swt.widgets.Display, java.lang.Runnable) line: 67	
	org.eclipse.jface.text.contentassist.CompletionProposalPopup.showProposals(boolean) line: 502	
	org.eclipse.jface.text.quickassist.QuickAssistAssistant$QuickAssistAssistantImpl(org.eclipse.jface.text.contentassist.ContentAssistant).showPossibleCompletions() line: 1874	
	org.eclipse.jdt.internal.ui.text.correction.JavaCorrectionAssistant(org.eclipse.jface.text.quickassist.QuickAssistAssistant).showPossibleQuickAssists() line: 113	
	org.eclipse.jdt.internal.ui.text.correction.JavaCorrectionAssistant.showPossibleQuickAssists() line: 194	
	org.eclipse.jdt.internal.ui.javaeditor.CompilationUnitEditor$AdaptedSourceViewer.doOperation(int) line: 201	
	org.eclipse.ui.texteditor.TextOperationAction.lambda$0() line: 130	
	org.eclipse.ui.texteditor.TextOperationAction$$Lambda.0x00000173d19dcb98.run() line: not available	
	org.eclipse.swt.custom.BusyIndicator.showWhile(org.eclipse.swt.widgets.Display, java.lang.Runnable) line: 67	
	org.eclipse.ui.texteditor.TextOperationAction.run() line: 130	
	org.eclipse.ui.texteditor.TextOperationAction(org.eclipse.jface.action.Action).runWithEvent(org.eclipse.swt.widgets.Event) line: 474	
	org.eclipse.jface.commands.ActionHandler.execute(org.eclipse.core.commands.ExecutionEvent) line: 124	
	org.eclipse.ui.internal.handlers.E4HandlerProxy.execute(org.eclipse.e4.core.contexts.IEclipseContext, java.util.Map, org.eclipse.swt.widgets.Event, org.eclipse.core.expressions.IEvaluationContext) line: 98	
	java.lang.invoke.LambdaForm$DMH.0x00000173d1a10400.invokeVirtual(java.lang.Object, java.lang.Object, java.lang.Object, java.lang.Object, java.lang.Object, java.lang.Object) line: not available	
	java.lang.invoke.LambdaForm$MH.0x00000173d1a11c00.invoke(java.lang.Object, java.lang.Object, java.lang.Object) line: not available	
	java.lang.invoke.Invokers$Holder.invokeExact_MT(java.lang.Object, java.lang.Object, java.lang.Object, java.lang.Object) line: not available	
	jdk.internal.reflect.DirectMethodHandleAccessor.invokeImpl(java.lang.Object, java.lang.Object[]) line: 157	
	jdk.internal.reflect.DirectMethodHandleAccessor.invoke(java.lang.Object, java.lang.Object[]) line: 103	
	java.lang.reflect.Method.invoke(java.lang.Object, java.lang.Object...) line: 580	
	org.eclipse.e4.core.internal.di.MethodRequestor.execute() line: 58	
	org.eclipse.e4.core.internal.di.InjectorImpl.invokeUsingClass(java.lang.Object, java.lang.Class<?>, java.lang.Class<? extends java.lang.annotation.Annotation>, java.lang.Object, org.eclipse.e4.core.di.suppliers.PrimaryObjectSupplier, org.eclipse.e4.core.di.suppliers.PrimaryObjectSupplier, boolean, boolean, boolean) line: 298	
	org.eclipse.e4.core.internal.di.InjectorImpl.invoke(java.lang.Object, java.lang.Class<? extends java.lang.annotation.Annotation>, java.lang.Object, org.eclipse.e4.core.di.suppliers.PrimaryObjectSupplier, org.eclipse.e4.core.di.suppliers.PrimaryObjectSupplier) line: 232	
	org.eclipse.e4.core.contexts.ContextInjectionFactory.invoke(java.lang.Object, java.lang.Class<? extends java.lang.annotation.Annotation>, org.eclipse.e4.core.contexts.IEclipseContext, org.eclipse.e4.core.contexts.IEclipseContext, java.lang.Object) line: 174	
	org.eclipse.ui.internal.WorkbenchHandlerServiceHandler(org.eclipse.e4.core.commands.internal.HandlerServiceHandler).execute(org.eclipse.core.commands.ExecutionEvent) line: 165	
	org.eclipse.core.commands.Command.executeWithChecks(org.eclipse.core.commands.ExecutionEvent) line: 488	
	org.eclipse.core.commands.ParameterizedCommand.executeWithChecks(java.lang.Object, java.lang.Object) line: 485	
	org.eclipse.e4.core.commands.internal.HandlerServiceImpl.executeHandler(org.eclipse.core.commands.ParameterizedCommand, org.eclipse.e4.core.contexts.IEclipseContext) line: 204	
	org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.executeCommand(org.eclipse.core.commands.ParameterizedCommand, org.eclipse.swt.widgets.Event) line: 308	
	org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.press(java.util.List<org.eclipse.jface.bindings.keys.KeyStroke>, org.eclipse.swt.widgets.Event) line: 569	
	org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.processKeyEvent(java.util.List<org.eclipse.jface.bindings.keys.KeyStroke>, org.eclipse.swt.widgets.Event) line: 644	
	org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.filterKeySequenceBindings(org.eclipse.swt.widgets.Event) line: 439	
	org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher$KeyDownFilter.handleEvent(org.eclipse.swt.widgets.Event) line: 96	
	org.eclipse.swt.widgets.EventTable.sendEvent(org.eclipse.swt.widgets.Event) line: 89	
	org.eclipse.swt.widgets.Display.filterEvent(org.eclipse.swt.widgets.Event) line: 1287	
	org.eclipse.swt.custom.StyledText(org.eclipse.swt.widgets.Widget).sendEvent(org.eclipse.swt.widgets.Event) line: 1065	
	org.eclipse.swt.custom.StyledText(org.eclipse.swt.widgets.Widget).sendEvent(int, org.eclipse.swt.widgets.Event, boolean) line: 1090	
	org.eclipse.swt.custom.StyledText(org.eclipse.swt.widgets.Widget).sendEvent(int, org.eclipse.swt.widgets.Event) line: 1075	
	org.eclipse.swt.custom.StyledText(org.eclipse.swt.widgets.Widget).sendKeyEvent(int, int, long, long, org.eclipse.swt.widgets.Event) line: 1117	
	org.eclipse.swt.custom.StyledText(org.eclipse.swt.widgets.Widget).sendKeyEvent(int, int, long, long) line: 1113	
	org.eclipse.swt.custom.StyledText(org.eclipse.swt.widgets.Widget).wmKeyDown(long, long, long) line: 1840	
	org.eclipse.swt.custom.StyledText(org.eclipse.swt.widgets.Control).WM_KEYDOWN(long, long) line: 5153	
	org.eclipse.swt.custom.StyledText(org.eclipse.swt.widgets.Canvas).WM_KEYDOWN(long, long) line: 406	
	org.eclipse.swt.custom.StyledText(org.eclipse.swt.widgets.Control).windowProc(long, int, long, long) line: 4776	
	org.eclipse.swt.custom.StyledText(org.eclipse.swt.widgets.Canvas).windowProc(long, int, long, long) line: 340	
	org.eclipse.swt.widgets.Display.windowProc(long, long, long, long) line: 5039	
	org.eclipse.swt.internal.win32.OS.DispatchMessage(org.eclipse.swt.internal.win32.MSG) line: not available [native method]	
	org.eclipse.swt.widgets.Display.readAndDispatch() line: 3658	
	org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run() line: 1151	
	org.eclipse.core.databinding.observable.Realm.runWithDefault(org.eclipse.core.databinding.observable.Realm, java.lang.Runnable) line: 339	
	org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(org.eclipse.e4.ui.model.application.MApplicationElement, org.eclipse.e4.core.contexts.IEclipseContext) line: 1042	
	org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(org.eclipse.e4.ui.model.application.MApplicationElement) line: 152	
	org.eclipse.ui.internal.Workbench.lambda$3(org.eclipse.swt.widgets.Display, org.eclipse.ui.application.WorkbenchAdvisor, int[]) line: 639	
	org.eclipse.ui.internal.Workbench$$Lambda.0x00000173d11c7968.run() line: not available	
	org.eclipse.core.databinding.observable.Realm.runWithDefault(org.eclipse.core.databinding.observable.Realm, java.lang.Runnable) line: 339	
	org.eclipse.ui.internal.Workbench.createAndRunWorkbench(org.eclipse.swt.widgets.Display, org.eclipse.ui.application.WorkbenchAdvisor) line: 546	
	org.eclipse.ui.PlatformUI.createAndRunWorkbench(org.eclipse.swt.widgets.Display, org.eclipse.ui.application.WorkbenchAdvisor) line: 173	
	org.eclipse.ui.internal.ide.application.IDEApplication.start(org.eclipse.equinox.app.IApplicationContext) line: 152	
	org.eclipse.equinox.internal.app.EclipseAppHandle.run(java.lang.Object) line: 208	
	org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(java.lang.Object) line: 143	
	org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(java.lang.Object) line: 109	
	org.eclipse.core.runtime.adaptor.EclipseStarter.run(java.lang.Object) line: 439	
	org.eclipse.core.runtime.adaptor.EclipseStarter.run(java.lang.String[], java.lang.Runnable) line: 271	
	java.lang.invoke.DirectMethodHandle$Holder.invokeStatic(java.lang.Object, java.lang.Object, java.lang.Object) line: not available	
	java.lang.invoke.LambdaForm$MH.0x00000173d1005c00.invoke(java.lang.Object, java.lang.Object, java.lang.Object, java.lang.Object) line: not available	
	java.lang.invoke.LambdaForm$MH.0x00000173d1006000.invokeExact_MT(java.lang.Object, java.lang.Object, java.lang.Object, java.lang.Object, java.lang.Object) line: not available	
	jdk.internal.reflect.DirectMethodHandleAccessor.invokeImpl(java.lang.Object, java.lang.Object[]) line: 155	
	jdk.internal.reflect.DirectMethodHandleAccessor.invoke(java.lang.Object, java.lang.Object[]) line: 103	
	java.lang.reflect.Method.invoke(java.lang.Object, java.lang.Object...) line: 580	
	org.eclipse.equinox.launcher.Main.invokeFramework(java.lang.String[], java.net.URL[]) line: 651	
	org.eclipse.equinox.launcher.Main.basicRun(java.lang.String[]) line: 588	
	org.eclipse.equinox.launcher.Main.run(java.lang.String[]) line: 1459	
	org.eclipse.equinox.launcher.Main.main(java.lang.String[]) line: 1432	

```

**Comment by jukzi** [comment]
@jjohnstn @chixiaye i don't have a clue what this code is all about, but it feels inappropriate that a "Quickfix" takes more then 3sec

My observation:
There are nested loops in 

-  ExtractTempRefactoring.processSelectedExpression() ,
-  ExtractTempRefactoring.evalStartAndEnd(IASTFragment[], int, Integer) ,
-  ChangedValueChecker.hasConflict()

totally calling node.accept(UpdateVisitor) 15 times but almost sequential.
only some calls to node.accept(UpdateVisitor) are slow.

**Comment by jjohnstn** [comment]
@jukzi This is partly caused by my latest change to not offer invalid extract to variable proposals.  The logic in question is verifying that there isn't a conflict created by moving the selected expression in front of the current statement (e.g. moving a call that will access a field that we are checking for null in a conditional).  There is a ThreadPool used to determine if a conflict occurs and it uses an awaitTermination call that has an upper limit of 5 seconds which is definitely part of the problem.   Maybe it would be better in some cases to simply run the visits directly if the number of statements is small.  The logic that uses the ThreadPool gets run when creating the actual change and happens when previewing or if the user chooses the proposal.  I had noticed recently there was a noticeable lag in the preview.  If the change cannot be made, the internal code bails with a status error and the edit formed ends up being invalid (e.g. the conditional you noticed that was removing the expression).  In the refactoring case, it catches this and does nothing.  I added code in QuickAssistProcessor to call checkFinalConditions() to create the change and determine it it was valid so the proposal would not be made in the first place.  Making the check code faster would definitely help, but I could do a couple things as well.  1. remove the unconditional checkFinalConditions() call and have an aborted change create a dummy edit which inserts nothing so clicking on it won't destroy the user's code 2. replace the preview with a message like some other refactoring proposals do and say something like (extract to temp if possible).  This should fix the problem and the work won't occur unless the user selects the change.
